src/integer.rs

```diff
@@ -116,44 +108,26 @@ impl<'env, K> IntegerWriter<'env, K>
 where
     K: PrimitiveInt,
 {
-    pub fn get<'s>(&'s self, k: K) -> Result<Option<Value<'s>>, StoreError> {
-        self.inner.get(Key::new(k)?)
-    }
-
-    pub fn put<'s>(&'s mut self, k: K, v: &Value) -> Result<(), StoreError> {
-        self.inner.put(Key::new(k)?, v)
-    }
-
-    fn abort(self) {
-        self.inner.abort();
+    pub(crate) fn new(writer: Writer<Key<K>>) -> IntegerWriter<K> {
+        IntegerWriter {
+            inner: writer,
+        }
     }
-}
 
-impl<K> IntegerStore<K>
```

Comment:
_Rkv::create_or_open_integer()_ still needs to return this type, and IntegerWriter/IntegerReader needs to require it. Otherwise, it'll be possible to retrieve a store with _Rkv::create_or_open_integer()_ and then write to it with the non-integer Writer:

```rust
    #[test]
    fn test_integer_store_with_non_integer_keys() {
        let root = Builder::new().prefix("test_integer_store_with_non_integer_keys").tempdir().expect("tempdir");
        fs::create_dir_all(root.path()).expect("dir created");
        let k = Rkv::new(root.path()).expect("new succeeded");
        let s = k.open_or_create_integer("s").expect("open");

        let mut writer = k.write().expect("writer");

        writer.put(&s, "123", &Value::Str("hello!")).expect("write");
        assert_eq!(writer.get(&s, "123").expect("read"), Some(Value::Str("hello!")));
    }
```

Which compiles and runs but doesn't retrieve the written value:

```
> > cargo test
> …
> test integer::tests::test_integer_store_with_non_integer_keys ... FAILED
> …
> failures:
> 
> ---- integer::tests::test_integer_store_with_non_integer_keys stdout ----
> 	thread 'integer::tests::test_integer_store_with_non_integer_keys' panicked at 'assertion failed: `(left == right)`
>   left: `None`,

```

Conversely, it'll be possible to retrieve a store with _Rkv::create_or_open()_ and then write to it with IntegerWriter, which compiles, runs, and retrieves the written value but probably isn't doing what the developer expects (i.e. applying LMDB's performance optimizations for integer-key stores).

The IntegerStore type enforces at compile time that all instances of IntegerWriter/IntegerReader are operating on a store that the developer retrieved via _Rkv::create_or_open_integer()_. I guess that still isn't entirely foolproof, as it's presumably still possible for the store in question to be a non-integer store (if it was previously created as such). But it's the equivalent of the current implementation.

(Ideally, I'd like Writer/Reader to be able to operate on both integer and non-integer stores, enforcing both of their key type constraints at compile time. But I'm not quite sure how to implement that. And in any case that's a separable problem. As you noted, IntegerStore is a WIP!)

Reply:
Nice catch, will take another look at this.

> Ideally, I'd like Writer/Reader to be able to operate on both integer and non-integer stores, enforcing both of their key type constraints at compile time.

I was also thinking about that possibility in the process of this code refactoring. Having one reader/writer to take care of both types would be great, but my previous experience of dealing with LMDB integer store in Python and Go told me it'd be hard to implement. Now let's see if we can figure this out in Rust.